src/basic/string-table.h

```diff
@@ -59,15 +59,19 @@ ssize_t string_table_lookup(const char * const *table, size_t len, const char *k
         scope type name##_from_string(const char *s) {                  \
                 unsigned u = 0;                                         \
                 type i;                                                 \
+                int r;                                                  \
                 if (!s)                                                 \
-                        return (type) -1;                               \
+                        return -EINVAL;                                 \
                 i = (type) string_table_lookup(name##_table, ELEMENTSOF(name##_table), s); \
                 if (i >= 0)                                             \
                         return i;                                       \
-                if (safe_atou(s, &u) >= 0 && u <= max)                  \
-                        return (type) u;                                \
-                return (type) -1;                                       \
-        }                                                               \
+                r = safe_atou(s, &u);                                   \
+                if (r < 0)                                              \
+                        return r;                                       \
+                if (u > max)                                            \
+                        return -ERANGE;                                 \
```

Comment:
hmm, i think we should return EINVAL in these two branches too, instead of ERANGE and whatever safe_atou() is. With defining the _INVALID field to -EINVAL we have the guarantee that it will fit into any enum this deals with, but we can't say the same for all other errnos, and they are beyond CHAR_MAX, hence I'd be careful.

I think we lose close to nothing if we just return EINVAL if the number is out of range, too, hence I'd just make this simple and declare that EINVAL is the only thing we can return, and it just generically means: input was not valid

Reply:
(alternative would be to fix all enum definitions to have another enum value in each enum definition, mapped to -ERRNO_MAX, but that force the size of each enum to be at least 12 bitm and i think is undesirable)

Reply:
Well, it's not *any* errno value, just two well-known values, 22 and 34. But OK, I'll change it to EINVAL.